### PR TITLE
gl: Support depth texel transformations using compute shaders

### PR DESCRIPTION
**Pull request by kd-11**
Implements texel transformation required for upload and download of depth formats. The memory ordering and endianness of PS3 depth formats makes it so that it does not map correctly to any OpenGL pack/unpack options and manual intervention is required. Also adds support for depth float transport bringing things up to par with vulkan when it comes to general feature support.

Implements https://github.com/RPCS3/rpcs3/issues/8714

### CODE REVIEW

**Comment by AniLeo** [comment]
`
Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz | 8 Threads | 15.52 GiB RAM | TSC: 2.592GHz | AVX+ | FMA3 | TSX-FA
`

`
POSIX, Name: Linux, Release: 5.8.6-1-MANJARO
`

Using mesa-git 20.3.0_devel.127615.83139aca59a-1 with the iris driver on Intel HD 530

---

**Yakuza 3 Demo** - Hangs when starting a new game
`F {RSX [0x0448220]} SIG: Thread terminated due to fatal error: Unsupported format`
image

**Persona 5** - Works, tested School Building and the end-game last boss map
image

**Comment by xddxd** [comment]
GTA IV just fatals with `F {RSX [0x008ef90]} SIG: Thread terminated due to fatal error: Unsupported format`
Log: [RPCS3.zip](https://github.com/RPCS3/rpcs3/files/5179141/RPCS3.zip)

It works on master, but *very* poorly and spams `RSX: Undefined format class for depth texture is not allowed`
Log: [master.log.gz](https://github.com/RPCS3/rpcs3/files/5179145/master.log.gz)
image

Windows 10 2004
R7 2700
RX 580 20.8.3

File: rpcs3/Emu/RSX/GL/GLTexture.cpp [hunk 1]
```diff
@@ -446,6 +454,166 @@ namespace gl
...
+		switch (pack_info.size)
+		{
+		case 1:
+			return nullptr;
+		case 2:
+			return get_compute_task<gl::cs_shuffle_16>();
+			break;
+		case 4:
+			return get_compute_task<gl::cs_shuffle_32>();
+			break;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLTexture.cpp:473]: break here and above ?

**Comment by xddxd** [comment]
GTA V fatals with `F 0:01:11.570702 {RSX [0x01166cc]} SIG: Thread terminated due to fatal error: Unsupported format`, while on master I can't even go in-game because the GL driver would just crash.

Windows 10 2004
R7 2700
RX 580 20.8.3

File: rpcs3/Emu/RSX/VK/VKTexture.cpp [hunk 2]
```diff
@@ -108,12 +108,12 @@ namespace vk
 			// 3. Do conversion with byteswap [D32->D16F]
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKTexture.cpp:108]: D32F ?

**Comment by kd-11** [comment]
Unsupported format error should be fixed now. Retesting is needed. GTA4 looks good now on OpenGL:
image

**Comment by AniLeo** [comment]
Not compiling

https://dev.azure.com/nekotekina/nekotekina/_build/results?buildId=3426&view=logs&j=f96cee46-76df-5e9b-533e-3bbb32f028c2&t=8fe0a607-2451-5d21-4bf8-7fbc14996860&l=2259

```
../rpcs3/Emu/RSX/GL/GLTexture.cpp:752:32: error: no matching function for call to 'max'
    compute_scratch_mem.create(std::max(512ull, staging_buffer.size() * 3), nullptr, buffer::memory_type::local, 0x88E6);
```

**Comment by xddxd** [comment]
TLoU spams `RSX: Undefined format class for depth texture is not allowed`
Screenshot_10
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/5179688/RPCS3.log.gz)
can add rsx capture if needed.
Master:
image
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/5179693/RPCS3.log.gz)

**Comment by xddxd** [comment]
RDR doesn't seem to differ visually from master on this PR, though it spams `RSX: Undefined format class for depth texture is not allowed`
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/5179701/RPCS3.log.gz)

**Comment by kd-11** [comment]
Fixed a bunch of problems, retest.

**Comment by MarioSonic2987** [comment]
GTA IV has now fixed graphics in OpenGL, same as Vulkan.
image
image

**Comment by AniLeo** [comment]
It's working fine on iris now for Yakuza 3 Demo

**Comment by woj1993** [comment]
How to download it? I wanted to test Sly Cooper.

**Comment by kd-11** [comment]
Swizzle/deswizzle will be handled in a separate PR.

**Comment by CoolEmuGuy** [comment]
I know this is not the place. But the regression for dynasty warriors gundam reborn since version 0.0.10-10468 hasn't been fixed yet.

**Pull request merged**